File: packages/core/android/src/main/java/com/posedetection/export/PoseExport.kt
package com.posedetection.export

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.RectF
import android.net.Uri
import com.google.mediapipe.framework.image.BitmapImageBuilder
import com.google.mediapipe.tasks.vision.poselandmarker.PoseLandmarkerResult
import com.posedetection.Skeleton
import com.posedetection.detector.PoseDetector
import com.posedetection.detector.StaticDetection
import com.posedetection.view.ContentFit
import com.posedetection.view.OverlayProjection
import com.posedetection.view.OverlayRenderer
import java.io.File
import java.io.FileOutputStream
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Paints the skeleton into a copy of a picked image or video and writes it into the app's sandbox.
 *
 * **Nothing here is allowed to slow the live camera down.** That is the whole shape of this file,
 * not a footnote to it, and it is bought four ways:
 *
 * 1. **Its own detector.** Never the camera's landmarker. Built here, used here, closed here.
 * 2. **CPU inference, always.** [PoseDetector.createForStillInput] asks for the CPU delegate, so an
 *    export cannot contend with the camera for the GPU its own inference is running on. Slower per
 *    frame, and entirely out of the way, which is the trade this package wants: an export has no
 *    deadline and a preview has one thirty times a second. The video pixel path does use the
 *    hardware codec, because that is the platform's cheap path and the alternative, converting
 *    every frame in Kotlin, would compete for far more CPU than the codec ever does for GPU.
 * 3. **A single background thread at low priority.** Serial, so two exports queue up behind each
 *    other instead of ganging up on the camera, and below the camera's own threads so the
 *    scheduler starves the export first.
 * 4. **Bounded memory.** One frame decoded at a time, one buffer encoded at a time, nothing
 *    accumulated. A long video costs what a short one costs, which is what keeps an export from
 *    ending as a memory-pressure kill of the camera it was running beside.
 *
 * Every path closes what it opened in a `finally`, so a throw, a cancel and a clean finish all
 * unwind the same way.
 */
internal object PoseExport {
    private val VIDEO_EXTENSIONS =
        setOf("mp4", "mov", "m4v", "3gp", "avi", "mkv", "webm")

    /** Serial and below the camera. See the note above; this is rule 3 and rule 3 is why it is here. */
    val executor =
        Executors.newSingleThreadExecutor { runnable ->
            Thread(runnable, "pose-export").apply {
                priority = Thread.MIN_PRIORITY
                isDaemon = true
            }
        }

    private val cancelled = ConcurrentHashMap<Int, AtomicBoolean>()

    fun cancel(taskId: Int) {
        cancelled[taskId]?.set(true)
    }

    fun isVideo(uri: String): Boolean {
        val path = Uri.parse(uri).path ?: uri
        return VIDEO_EXTENSIONS.contains(path.substringAfterLast('.', "").lowercase())
    }

    fun run(
        context: Context,
        uri: String,
        raw: Map<*, *>?,
        taskId: Int,
        onProgress: (Float) -> Unit,
    ): ExportSummary {
        val sourceName =
            (Uri.parse(uri).lastPathSegment ?: "pose")
                .substringAfterLast('/')
                .substringBeforeLast('.')
        val options = ExportOptions.parse(context, raw, sourceName)

        // An entry exists only while the job runs, so cancelling a task nobody started is a no-op
        // rather than a note kept for the life of the process.
        val flag = AtomicBoolean(false)
        cancelled[taskId] = flag
        try {
            return if (isVideo(uri)) {
                VideoExporter(context, uri, options, flag, onProgress).run()
            } else {
                exportImage(context, uri, options, onProgress)
            }
        } finally {
            cancelled.remove(taskId)
        }
    }

    // MARK: Stills

    private fun exportImage(
        context: Context,
        uri: String,
        options: ExportOptions,
        onProgress: (Float) -> Unit,
    ): ExportSummary {
        val source =
            StaticDetection.loadBitmap(context, uri)
                ?: throw ExportError("could not read an image from $uri")

        var detector: PoseDetector? = null
        val result: PoseLandmarkerResult
        try {
            detector =
                PoseDetector.createForStillInput(
                    context,
                    StaticDetection.requireModel(context),
                    options.maxPoses,
                    video = false,
                )
            result = detector.detectImage(BitmapImageBuilder(source).build())
        } finally {
            detector?.close()
        }
        onProgress(0.6f)

        val canvas = ExportCanvas.size(source.width, source.height, options.maxSize)
        val painted = Bitmap.createBitmap(canvas[0], canvas[1], Bitmap.Config.ARGB_8888)
        try {
            paint(painted, source, result, options)
            FileOutputStream(File(options.directory, "${options.fileName}.jpg")).use { stream ->
                painted.compress(Bitmap.CompressFormat.JPEG, options.quality, stream)
            }
        } finally {
            painted.recycle()
            source.recycle()
        }
        onProgress(1f)

        return ExportSummary(
            file = File(options.directory, "${options.fileName}.jpg"),
            width = canvas[0],
            height = canvas[1],
            durationMs = 0,
            frameCount = 1,
            posesFound = result.landmarks().size,
        )
    }

    private fun paint(
        target: Bitmap,
        source: Bitmap,
        result: PoseLandmarkerResult,
        options: ExportOptions,
    ) {
        // Fit, not fill: cropping a picture the user picked would cut away part of the very thing
        // they asked to have painted.
        val projection =
            OverlayProjection(
                source.width,
                source.height,
                target.width.toFloat(),
                target.height.toFloat(),
                ContentFit.FIT,
            )
        val canvas = Canvas(target)
        canvas.drawColor(Color.BLACK)
        // The rect is built here rather than on the projection: that class is deliberately free of
        // `android.graphics` so it can run under a plain JVM test.
        canvas.drawBitmap(source, null, projection.rect(), null)

        val landmarks = FloatArray(Skeleton.LANDMARK_COUNT * Skeleton.LANDMARK_STRIDE)
        if (!options.drawOverlay || !fill(landmarks, result)) return

        val renderer = OverlayRenderer(ExportCanvas.overlayScale(target.width, target.height))
        renderer.config = options.overlay
        renderer.draw(
            canvas,
            landmarks,
            projection,
            // A file is never mirrored: what was picked is what gets painted.
            mirrored = false,
            sourceWidth = source.width,
            sourceHeight = source.height,
        )
    }

    private fun OverlayProjection.rect(): RectF = RectF(left, top, left + width, top + height)

    /** The primary pose as the flat buffer the renderer and the geometry both read. */
    fun fill(
        landmarks: FloatArray,
        result: PoseLandmarkerResult,
    ): Boolean {
        val pose = result.landmarks().firstOrNull() ?: return false
        val count = minOf(Skeleton.LANDMARK_COUNT, pose.size)
        for (index in 0 until count) {
            val point = pose[index]
            val base = index * Skeleton.LANDMARK_STRIDE
            landmarks[base + Skeleton.OFFSET_X] = point.x()
            landmarks[base + Skeleton.OFFSET_Y] = point.y()
            landmarks[base + Skeleton.OFFSET_Z] = point.z()
            landmarks[base + Skeleton.OFFSET_VISIBILITY] =
                if (point.visibility().isPresent) point.visibility().get() else 0f
        }
        return true
    }
}

File: packages/core/ios/export/PoseExport.swift
import AVFoundation
import MediaPipeTasksVision
import UIKit

/**
 Paints the skeleton into a copy of a picked image or video and writes it into the app's sandbox.

 **Nothing here is allowed to slow the live camera down.** That is the whole shape of this file,
 not a footnote to it, and it is bought four ways:

 1. **Its own detector.** Never the camera's landmarker. Built here, used here, released here.
 2. **CPU inference, always.** `createForStillInput` asks for the CPU delegate, so an export cannot
    contend with the camera for the GPU that its own inference is running on. Slower per frame, and
    entirely out of the way, which is the trade this package wants: an export has no deadline and
    a preview has one thirty times a second.
 3. **A `.utility` serial queue.** Below the camera's `.userInitiated` analysis queue, so under load
    the scheduler starves the export rather than sharing evenly. Serial, so two exports queue up
    behind each other instead of ganging up on the camera.
 4. **Bounded memory.** One frame decoded at a time, one pooled buffer encoded at a time, nothing
    accumulated across frames. A long video costs the same as a short one, which is what keeps an
    export from ending as a memory-pressure kill of the camera it was running beside.

 Every path releases on the way out through `defer`, so a throw, a cancel and a clean finish all
 unwind the same way.
 */
enum PoseExport {
  private static let videoExtensions: Set<String> = ["mp4", "mov", "m4v", "3gp", "avi", "mkv", "webm"]

  /// Serial and below the camera. See the note above; this is rule 3 and rule 3 is why it is here.
  static let queue = DispatchQueue(label: "com.posedetection.export", qos: .utility)

  private static let running = CancelRegistry()

  static func cancel(taskId: Int) {
    running.cancel(taskId)
  }

  static func isVideo(uri: String) -> Bool {
    let ext = (URL(string: uri) ?? URL(fileURLWithPath: uri)).pathExtension.lowercased()
    return videoExtensions.contains(ext)
  }

  static func resolve(uri: String) throws -> URL {
    guard let url = URL(string: uri) ?? URL(string: "file://\(uri)") else {
      throw ExportError("could not read a file from \(uri)")
    }
    return url
  }

  static func run(
    uri: String,
    raw: [String: Any]?,
    taskId: Int,
    onProgress: @escaping (Float) -> Void
  ) throws -> ExportSummary {
    let source = try resolve(uri: uri)
    let options = try ExportOptions.parse(raw, sourceName: source.deletingPathExtension().lastPathComponent)
    // The two values that decide who gets painted, and the pair most worth seeing when an export
    // finds fewer bodies than expected: `minConfidence` follows `maxPoses` unless the caller set it,
    // and the resolved number is not visible from JavaScript otherwise.
    PoseLog.info(.engine, "export maxPoses=\(options.maxPoses) minConfidence=\(options.minConfidence)")

    running.begin(taskId)
    // Suspension is the one failure the cleanup blocks cannot see: the threads freeze mid-write,
    // and if the system then terminates the app nothing unwinds. The background task buys the
    // minutes an export actually needs, and its expiry cancels this task so the pipeline leaves
    // through the same cleanup as every other failure.
    let background = UIApplication.shared.beginBackgroundTask(withName: "pose-export") {
      running.cancel(taskId)
    }
    defer {
      running.end(taskId)
      if background != .invalid {
        UIApplication.shared.endBackgroundTask(background)
      }
    }

    if isVideo(uri: uri) {
      let exporter = VideoExporter(
        source: source,
        options: options,
        isCancelled: { running.isCancelled(taskId) },
        onProgress: onProgress
      )
      return try exporter.run()
    }
    return try exportImage(source: source, options: options, onProgress: onProgress)
  }

  // MARK: - Stills

  private static func exportImage(
    source: URL,
    options: ExportOptions,
    onProgress: @escaping (Float) -> Void
  ) throws -> ExportSummary {
    guard let image = StaticDetection.loadImage(uri: source.absoluteString) else {
      throw ExportError("could not read an image from \(source.lastPathComponent)")
    }

    let detector = try PoseDetector.createForStillInput(
      modelPath: try StaticDetection.requireModel(),
      maxPoses: options.maxPoses,
      minConfidence: options.minConfidence,
      video: false
    )
    let result = try detector.detectImage(try MPImage(uiImage: image))
    onProgress(0.6)

    let display = CGSize(width: image.size.width * image.scale, height: image.size.height * image.scale)
    let canvas = exportCanvasSize(display: display, maxSize: options.maxSize)
    // Fit, not fill: cropping a picture the user picked would cut away part of the very thing they
    // asked to have painted.
    let projection = OverlayProjection(
      source: display,
      bounds: CGRect(origin: .zero, size: canvas),
      fit: .fit
    )

    let painted = paint(
      image,
      result: result,
      options: options,
      projection: projection,
      geometry: (display: display, canvas: canvas)
    )

    let url = options.directory.appendingPathComponent("\(options.fileName).jpg")
    guard let data = painted.jpegData(compressionQuality: options.quality) else {
      throw ExportError("could not encode the painted image")
    }
    try data.write(to: url, options: .atomic)
    onProgress(1)

    return ExportSummary(
      url: url,
      width: Int(canvas.width),
      height: Int(canvas.height),
      durationMs: 0,
      frameCount: 1,
      posesFound: result.landmarks.count
    )
  }

  /// The picture with every detected skeleton on top of it.
  private static func paint(
    _ image: UIImage,
    result: PoseLandmarkerResult,
    options: ExportOptions,
    projection: OverlayProjection,
    geometry: (display: CGSize, canvas: CGSize)
  ) -> UIImage {
    let canvas = geometry.canvas
    let format = UIGraphicsImageRendererFormat()
    format.scale = 1
    format.opaque = true
    let scale = overlayScale(canvas: canvas)
    let palette = OverlayPalette(options.overlay, scale: scale)

    // Safe off the main thread, unlike the UIGraphicsBeginImageContext family it replaced, which is
    // the reason an export can render at all without hopping onto the thread the camera draws on.
    return UIGraphicsImageRenderer(size: canvas, format: format).image { context in
      image.draw(in: projection.rect)
      guard options.drawOverlay else { return }
      for landmarks in poses(result) {
        var renderer = OverlayRenderer(
          config: options.overlay,
          palette: palette,
          landmarks: landmarks,
          projection: projection,
          // A file is never mirrored: what was picked is what gets painted.
          mirrored: false,
          sourceWidth: Int(geometry.display.width),
          sourceHeight: Int(geometry.display.height)
        )
        renderer.scale = scale
        renderer.draw(into: context.cgContext)
      }
    }
  }

  /**
   Every pose in the frame as flat buffers the renderer reads, empty when nobody was in it.

   All of them, not the front one: `maxPoses` is an export option, and painting one skeleton onto a
   frame the caller asked to have five detected in would silently ignore what they asked for.
   */
  static func poses(_ result: PoseLandmarkerResult) -> [[Float]] {
    return result.landmarks.map { pose in
      var landmarks = [Float](repeating: 0, count: Skeleton.landmarkCount * Skeleton.landmarkStride)
      for index in 0..<min(Skeleton.landmarkCount, pose.count) {
        let point = pose[index]
        let base = index * Skeleton.landmarkStride
        landmarks[base + Skeleton.offsetX] = point.x
        landmarks[base + Skeleton.offsetY] = point.y
        landmarks[base + Skeleton.offsetZ] = point.z
        landmarks[base + Skeleton.offsetVisibility] = point.visibility?.floatValue ?? 0
      }
      return landmarks
    }
  }
}
